controllers/bmcjob_controller.go

```diff
@@ -18,45 +18,249 @@ package controllers
 
 import (
 	"context"
+	"fmt"
 
-	"k8s.io/apimachinery/pkg/runtime"
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
 
 	bmcv1alpha1 "github.com/tinkerbell/rufio/api/v1alpha1"
 )
 
+// Index key for Job Owner Name
+const jobOwnerKey = ".metadata.controller"
+
 // BMCJobReconciler reconciles a BMCJob object
 type BMCJobReconciler struct {
-	client.Client
-	Scheme *runtime.Scheme
+	client client.Client
+	logger logr.Logger
+}
+
+// NewBMCJobReconciler returns a new BMCJobReconciler
+func NewBMCJobReconciler(client client.Client, logger logr.Logger) *BMCJobReconciler {
+	return &BMCJobReconciler{
+		client: client,
+		logger: logger,
+	}
 }
 
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the BMCJob object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+// Reconcile runs a BMCJob.
+// Creates the individual BMCTasks on the cluster.
+// Watches for BMCTask and creates next BMCJob Task based on conditions.
 func (r *BMCJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := r.logger.WithValues("BMCJob", req.NamespacedName)
+	logger.Info("Reconciling BMCJob")
+
+	// Fetch the bmcJob object
+	bmcJob := &bmcv1alpha1.BMCJob{}
+	err := r.client.Get(ctx, req.NamespacedName, bmcJob)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Failed to get BMCJob")
+		return ctrl.Result{}, err
+	}
+
+	// Deletion is a noop.
+	if !bmcJob.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	// Job is Completed or Failed is noop.
+	if bmcJob.HasCondition(bmcv1alpha1.JobCompleted, bmcv1alpha1.ConditionTrue) ||
+		bmcJob.HasCondition(bmcv1alpha1.JobFailed, bmcv1alpha1.ConditionTrue) {
```

Comment:
Why do we think `Failed` is a condition? Wouldn't that just be state on the `Status` (`Success` or `Failed` being the valid values)? In the event its `Failed` a description field of some sort. This way, consumers can watch for the condition "Completed" to become true before inspecting the actual result. 

I don't recall what we said about `Failed` and `Success` etc. I'd be OK with allowing this to pass and changing it in a separate PR (or stripping it out of this one if possible). There's a bunch of good stuff here already that we don't want to hold up.

We need to be careful that we're not encoding job state into conditions.